source/solvers/heat_transfer.cc

```diff
@@ -23,589 +27,237 @@
-  const Parameters::SimulationControl::TimeSteppingMethod time_stepping_method)
+  const Parameters::SimulationControl::
+    TimeSteppingMethod /*time_stepping_method*/)
 {
-  assemble_system<true>(time_stepping_method);
+  assemble_system_matrix();
+  assemble_system_rhs();
 }
 
 
 template <int dim>
 void
-HeatTransfer<dim>::assemble_rhs(
-  const Parameters::SimulationControl::TimeSteppingMethod time_stepping_method)
+HeatTransfer<dim>::assemble_rhs(const Parameters::SimulationControl::
+                                  TimeSteppingMethod /*time_stepping_method*/)
 {
-  assemble_system<false>(time_stepping_method);
+  assemble_system_rhs();
 }
 
-
 template <int dim>
-template <bool assemble_matrix>
 void
-HeatTransfer<dim>::assemble_system(
-  const Parameters::SimulationControl::TimeSteppingMethod time_stepping_method)
+HeatTransfer<dim>::setup_assemblers()
 {
-  auto &physical_properties = this->simulation_parameters.physical_properties;
-
-  // Gather physical properties in case of mono fluids simulations (to be
-  // modified by cell in case of multiple fluids simulations)
-  double density              = physical_properties.density;
-  double specific_heat        = physical_properties.specific_heat;
-  double thermal_conductivity = physical_properties.thermal_conductivity;
-  double viscosity            = physical_properties.viscosity;
-
-  double dynamic_viscosity = viscosity * density;
-  double rho_cp            = density * specific_heat;
-  double alpha             = thermal_conductivity / rho_cp;
-
-  if (assemble_matrix)
-    system_matrix = 0;
-  system_rhs = 0;
-
-  // Vector for the BDF coefficients
-  // The coefficients are stored in the following fashion :
-  // 0 - n+1
-  // 1 - n
-  // 2 - n-1
-  // 3 - n-2
-  std::vector<double> time_steps_vector =
-    simulation_control->get_time_steps_vector();
-
-  // Time steps and inverse time steps which is used for numerous calculations
-  const double dt  = time_steps_vector[0];
-  const double sdt = 1. / dt;
-
-
-  Vector<double> bdf_coefs;
-
-  if (time_stepping_method ==
-        Parameters::SimulationControl::TimeSteppingMethod::bdf1 ||
-      time_stepping_method ==
-        Parameters::SimulationControl::TimeSteppingMethod::steady_bdf)
-    bdf_coefs = bdf_coefficients(1, time_steps_vector);
-
-  if (time_stepping_method ==
-      Parameters::SimulationControl::TimeSteppingMethod::bdf2)
-    bdf_coefs = bdf_coefficients(2, time_steps_vector);
-
-  if (time_stepping_method ==
-      Parameters::SimulationControl::TimeSteppingMethod::bdf3)
-    bdf_coefs = bdf_coefficients(3, time_steps_vector);
-
-  if (time_stepping_method ==
-        Parameters::SimulationControl::TimeSteppingMethod::sdirk22_1 ||
-      time_stepping_method ==
-        Parameters::SimulationControl::TimeSteppingMethod::sdirk33_1)
+  this->assemblers.clear();
+
+  // Robin boundary condition
+  this->assemblers.push_back(std::make_shared<HeatTransferAssemblerRBC<dim>>(
```

Comment:
I would give this class a more explicit name. HeatTransferAssemblerRobinBC or HeatTransferAssemblerConvectionBC for example.